dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1052 @@
+  argument anyways, we make it where they either can here or wait until `GetResultAsync` to do so.
+  * We could have taken a proxy approach, but that would have made the code look like
+    `client.GetWorkflowHandle<GreetingWorkflow>("my wf id", wf => wf.RunAsync)` which is not much better and is hard to
+    get right with all of the overloads so you can have the lambda as the last param. This is what Azure Durable
+    Entities does.
+* Cancellation tokens (and RPC metadata and RPC timeout and such) will be in the call options, not as the last parameter
+  in a bunch of overloads like many other .NET methods
+* Older versions of .NET don't have `IAsyncEnumerator` and we don't want to go out of our way to provide a special
+  `ListWorkflows` for them
+  * Older versions can just use raw gRPC or not have this feature. These older .NET versions are really old and we don't
+    expect hardly any users on them.
+  * Note, `ListWorkflows` is not `ListWorkflowsAsync` because the first page is lazily fetched on first iteration
+  * `WorkflowExecutionEnumerator` has a `MapHistories` on it that returns `IAsyncEnumerator<WorkflowHistory>` (or is an
+    extension method)
+
+### WorkflowHandle
+
+Looks like:
+
+```csharp
+namespace Temporal.Client;
+
+public class WorkflowHandle<TResult>
+{
+    public string ID { get; }
+    public string? RunID { get; }
+    public string? ResultRunID { get; }
+    public string? FirstExecutionRunID { get; }
+
+    public async Task<TResult> GetResultAsync(
+        bool followRuns = true, RPCOptions? options = null) { /*...*/ }
+
+    public async Task<TResultLocal> GetResultAsync<TResultLocal>(
+        bool followRuns = true, RPCOptions? options = null) { /*...*/ }
+
+    public async Task SignalAsync(string signal, object[] args, RPCOptions? options = null) { /*...*/ }
+
+    public async Task SignalAsync(Func<Task> signal, RPCOptions? options = null) { /*...*/ }
+
+    public async Task SignalAsync<T>(Func<T, Task> signal, T arg, RPCOptions? options = null) { /*...*/ }
+
+    public async Task<TQueryResult> QueryAsync<TQueryResult>(
+      string query, object[] args, RPCOptions? options = null) { /*...*/ }
+
+    public async Task<TQueryResult> QueryAsync<TQueryResult>(
+      Func<TQueryResult> query, RPCOptions? options = null) { /*...*/ }
+
+    public async Task<TQueryResult> QueryAsync<T, TQueryResult>(
+      Func<T, TQueryResult> query, T arg, RPCOptions? options = null) { /*...*/ }
+
+    public async Task CancelAsync(RPCOptions? options = null) { /*...*/ }
+
+    public async Task TerminateAsync(
+        string reason? = null, object[]? args = null, RPCOptions? options = null) { /*...*/ }
+
+    public async Task<WorkflowExecutionDescription> DescribeAsync(RPCOptions? options = null) { /*...*/ }
+
+#if NETCOREAPP3_0_OR_GREATER
+
+    public async Task<WorkflowHistory> FetchHistoryAsync(
+        WorkflowHistoryEventFilterType eventFilterType = WorkflowHistoryEventFilterType.AllEvents,
```

Comment:
I default them. But I think they should be exposed. If we're allowed to wait on events and filter them so should a user.

Reply:
What bothers me is that they get in the way of `RpcOptions`, which means that a user who wants to set `RpcOptions` is forced to explicitly set all of these other (very rarely used) options. But maybe named arguments would help there, not sure how they interact with optional args. Otherwise, I'd suggest an overload on the signature with `(RpcOptions)` only.

Reply:
> But maybe named arguments would help there, not sure how they interact with optional args.

Right. You can ignore optional args and use named args. So you can have:

```csharp
await myHandle.FetchHistoryAsync(options: new(timeout=TimeSpan.FromMinutes(3)));
```

That reminds me I do need to change that arg name to `rpcOptions` instead of just `options`.